lib/classes/PluginManager.js

```diff
@@ -0,0 +1,120 @@
+'use strict';
+
+const Utils = require('./Utils');
+const path = require('path');
+const has = require('lodash').has;
+const forEach = require('lodash').forEach;
+
+class PluginManager {
+  constructor(serverless) {
+    this._serverless = serverless;
+    this._pluginInstances = [];
+    this._commandsList = [];
+    this._commands = {};
+  }
+
+  loadAllPlugins(servicePlugins) {
+    this._loadCorePlugins();
+    this._loadServicePlugins(servicePlugins);
+  }
+
+  runCommand(command) {
+    const events = this._getEvents(command.split(' '), this._commands);
+    // collect all relevant hooks
+    let hooks = [];
+    events.forEach((event) => {
+      const hooksForEvent = [];
+      this._pluginInstances.forEach((pluginInstance) => {
+        forEach(pluginInstance.hooks, (hook, hookKey) => {
+          // normalize the hookKey and event ("foo:beforeBar" becomes "foo:beforebar")
+          const normalizedHookKey = hookKey.substr(0, hookKey.lastIndexOf(':') + 1) +
+            hookKey.substr((hookKey.lastIndexOf(':')) + 1).toLowerCase();
+          const normalizedEvent = event.substr(0, event.lastIndexOf(':') + 1) +
+            event.substr((event.lastIndexOf(':')) + 1).toLowerCase();
+
+          if (normalizedHookKey === normalizedEvent) {
+            hooksForEvent.push(hook);
+          }
+        });
+      });
+      hooks = hooks.concat(hooksForEvent);
+    });
+
+    // run all relevant hooks one after another
+    hooks.forEach((hook) => {
+      const returnValue = hook();
+
+      // check if a Promise is returned
+      if (returnValue && returnValue.then instanceof Function) {
+        returnValue.then((value) => {
+          return value;
+        });
+      }
+    });
+  }
+
+  _addPlugin(Plugin) {
+    this._pluginInstances.push(new Plugin());
```

Comment:
I think pluginInstances can be public (e.g. we can just call it plugins). Imho its something that other plugins could be interested in so I wouldn't put _ in front of it.